### PR TITLE
Added typings

### PR DESCRIPTION
**Pull request by sharmapukar217**

### CODE REVIEW

**Review by jonnydgreen — CHANGES_REQUESTED** [review]
Nice one! :) I've done an initial pass. Could you also add some types tests (using `tsd`) as well? You can use `mercurius-js` as a reference for this:
 - Example command: https://github.com/mercurius-js/mercurius/blob/master/package.json#L15
 - Example types test file: https://github.com/mercurius-js/mercurius/blob/master/test/types/index.ts

**Review by mcollina — CHANGES_REQUESTED** [review]
Could you add tests with https://www.npmjs.com/package/tsd?

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { FastifyInstance } from "fastify";
+import { DirectiveNode, GraphQLResolveInfo } from "graphql";
+import { MercuriusContext } from "mercurius";
+
+interface MercuriusAuthOptions {
+  authContext?: (context: MercuriusContext) => object | Promise<object>;
+  applyPolicy: (
+    authDirectiveAST: DirectiveNode,
+    parent: object,
+    args: Record<string, any>,
```
[the remark this answers is not in the document]
Comment (sharmapukar217) [reply index.d.ts:10]: i thought { [argName: string]: any }  and Record<string, any> were similar

[the remark this answers is not in the document]
Comment (jonnydgreen) [reply index.d.ts:10]: You are correct, yep - this is also fine as a default (the `{ [argName: string]: any }` came from what `graphql-js` has used)

[the remark this answers is not in the document]
Comment (sharmapukar217) [reply index.d.ts:10]: oh ok following `graphql-js`

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import { DirectiveNode, GraphQLResolveInfo } from "graphql";
+import { MercuriusContext } from "mercurius";
+
+interface MercuriusAuthOptions {
+  authContext?: (context: MercuriusContext) => object | Promise<object>;
+  applyPolicy: (
+    authDirectiveAST: DirectiveNode,
+    parent: object,
+    args: Record<string, any>,
+    context: MercuriusContext,
```
[the remark this answers is not in the document]
Comment (sharmapukar217) [reply index.d.ts:11]: In order to accept generic, do we need to export a type for applyPolicy function ? I couldn't figure out other way to pass generic on function inside an object

[the remark this answers is not in the document]
Comment (jonnydgreen) [reply index.d.ts:11]: Yep, exporting an `applyPolicyHandler` type should do the trick here, then you can pass the generics all the way through from the options type!

**Pull request closed without merging**